services/transactionService.ts

```diff
@@ -0,0 +1,554 @@
+  const offset = input.offset || 1000
+  // https://docs.etherscan.io/api-endpoints/accounts#get-a-list-of-erc20-token-transfer-events-by-address
+  // https://blockscout.com/xdai/mainnet/api-docs#account
+  const result = await axios.get(getEtherscanOrBlockScoutUrl(networkId), {
+    params: {
+      module: 'account',
+      action: 'tokentx',
+      page,
+      offset,
+      address,
+      contractAddress,
+      sort: 'desc'
+    }
+  })
+  if (
+    !result.data ||
+    !result.data.result ||
+    !Array.isArray(result.data.result)
+  ) {
+    console.log(
+      'getListOfTokenTransferTransactionsByAddress transactions are empty',
+      {
+        networkId,
+        page,
+        offset,
+        address,
+        contractAddress,
+        response: result.data
+      }
+    )
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUNT_IN_USER_HISTORY)
+  }
+
+  return result.data.result
+    .filter(tx => {
+      return tx.from.toLowerCase() === input.address.toLowerCase()
+    })
+    .map(tx => {
+      return {
+        hash: tx.hash,
+        nonce: Number(tx.nonce),
+        amount: Number(tx.value) / 10 ** Number(tx.tokenDecimal),
+        from: tx.from,
+        to: tx.to,
+        currency: tx.tokenSymbol
+      }
+    })
+}
+
+export const checkIfTransactionHasBeenSpeedup = async (data: {
+  input: TransactionDetailInput
+  page?: number
+  contractAddress?: string
+}): Promise<NetworkTransactionInfo> => {
+  console.log('checkIfTransactionHasBeenSpeedup called', data)
+  const { input, page = 1, contractAddress } = data
+  const nonce = input.nonce as number
+  const userRecentTransactions = contractAddress
+    ? await getListOfTokenTransferTransactionsByAddress({
+        address: input.fromAddress,
+        page,
+        networkId: input.networkId,
+        contractAddress
+      })
+    : await getListOfNormalTransactionsByAddress({
+        address: input.fromAddress,
+        page,
+        networkId: input.networkId
+      })
+  if (userRecentTransactions.length === 0) {
+    console.log(
+      'checkIfTransactionHasBeenSpeedup, no more found donations for address',
+      {
+        page,
+        address: input.fromAddress,
+        contractAddress
+      }
+    )
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUND)
+  }
+  const foundTransaction = userRecentTransactions.find(
+    tx => tx.nonce === input.nonce
+  )
+  if (
+    foundTransaction &&
+    foundTransaction.to.toLowerCase() === input.toAddress.toLowerCase() &&
+    foundTransaction.amount === input.amount
+  ) {
+    return { ...foundTransaction, speedup: true }
+  }
+  console.log('checkIfTransactionHasBeenSpeedup', {
+    recentDonationsLength: userRecentTransactions.length,
+    lastDonation: userRecentTransactions[userRecentTransactions.length - 1],
+    nonce,
+    page
+  })
+  const smallestNonce: number = userRecentTransactions[
+    userRecentTransactions.length - 1
+  ].nonce as number
+
+  if (smallestNonce < nonce) {
```

Comment:
What if user tx has not mined yet?

Reply:
Then we don't change the transaction status and it would be checked in next cycle